openmdao/core/implicitcomponent.py

```diff
@@ -317,6 +356,258 @@ def add_output(self, name, val=1.0, **kwargs):
 
         return metadata
 
+    def add_residual(self, name, shape=(), units=None, desc='', ref=1.0):
+        """
+        Add an residual variable to the component.
+
+        Note that the total size of the residual vector must match the total size of
+        the outputs vector for this component.
+
+        Parameters
+        ----------
+        name : str
+            Name of the residual in this component's namespace.
+        shape : int or tuple
+            Shape of this residual.
+        units : str or None
+            Units in which this residual will be given to the user when requested.
```

Comment:
Same with 'units' -- what if there is a conflict between units declared in the output and the residual?